### PR TITLE
[WIP] Instance Encounters

### PR DESCRIPTION
**Pull request by aaron02**
**Todo / Checklist**

- [x] Add More Instances
- [x] Add Sql File
- [x] Script ICC as an Example for new Scripts, Goal should be to add new Funtctions, modify or delete obsolete Functions to give a good base at Instance Scripting // edit changes got to bit just scriptet to deathwhisper now more coming soon)

**Tests Performed:** 
    - Build
    - Tested in-game

Fixes: 
https://github.com/AscEmu/AscEmu/issues/174

### CODE REVIEW

**Comment by Zyres** [comment]
Welcome back @aaron02 , glad to see a PR from you.
I've worked on encounters a few years ago. There should already exist a table for it.
If you want, you can chat with me in discord if you need some help 🤗
Keeper up your work 🥳

**Comment by schnek** [comment]
there are a couple of ways to put the commits history in order
one of them program [desktop.github](https://desktop.github.com/)  (yes, I know that this can be written in the console)
image
another way to delete commits permanently [clear-history-and-merge-commits](http://board.ascemu.org/index.php/Thread/264-how-to-clear-history-and-merge-commits/?postID=929#post929)

🐛🕷😅

File: src/world/Spell/Customization/HackFixes.cpp [hunk 1]
```diff
@@ -1127,6 +1127,16 @@ void SpellMgr::applyHackFixes()
...
             } break;
+            case 69075: // Bone Storm (Lord Marrowgar)
+            case 70834: // Bone Storm (Lord Marrowgar)
+            case 70835: // Bone Storm (Lord Marrowgar)
+            case 70836: // Bone Storm (Lord Marrowgar)
+            case 72864: // Death Plague (Rotting Frost Giant)
+            case 71160: // Plague Stench (Stinky)
+            case 71161: // Plague Stench (Stinky)
+            case 71123: // Decimate (Stinky & Precious)
+                sp->setEffectRadiusIndex(sSpellRadiusStore.LookupEntry(12)->ID, 0); // 100yd
```
Comment (Appled) [inline src/world/Spell/Customization/HackFixes.cpp:1138]: DBC data does not seem to have radius for these spells so this is needed, but 100 yards seem quite large radius

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 2]
```diff
@@ -1216,6 +1178,71 @@ class BoneSlice : public SpellScript
...
+
+///////////////////////////////////////////////////////
+// Boss: Valithria Dreamwalker
+// ...
+//
+//
+//
+//
+
+///////////////////////////////////////////////////////
```
Comment (schnek) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1193]: //////////////////////////////////////////////////////////////////////////////////////////
// Boss: Gunship Battle Alliance

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 3]
```diff
@@ -1216,6 +1178,71 @@ class BoneSlice : public SpellScript
     uint32_t targetCount;
 };
 
+//////////////////////////////////////////////////////////////////////////////////////////
+// Boss: Lady Deathwhisper
+//////////////////////////////////////////////////////////////////////////////////////////
+
+///////////////////////////////////////////////////////
```
Comment (schnek) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1185]: //////////////////////////////////////////////////////////////////////////////////////////
// Boss: Valithria Dreamwalker

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 4]
```diff
@@ -1216,6 +1178,71 @@ class BoneSlice : public SpellScript
     uint32_t targetCount;
 };
 
+//////////////////////////////////////////////////////////////////////////////////////////
+// Boss: Lady Deathwhisper
+//////////////////////////////////////////////////////////////////////////////////////////
```
Comment (schnek) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1183]: //////////////////////////////////////////////////////////////////////////////////////////
// Boss: Lady Deathwhisper

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 5]
```diff
@@ -1171,6 +1129,10 @@ class ColdflameDamage: public SpellScript
     }
 };
 
+//////////////////////////////////////////////////////////////////////////////////////////
```
Comment (schnek) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1132]: //////////////////////////////////////////////////////////////////////////////////////////
// Spell: Bone Slice

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 6]
```diff
@@ -1130,6 +1084,10 @@ class ColdflameBonestorm : public SpellScript
...
```
Comment (schnek) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1087]: //////////////////////////////////////////////////////////////////////////////////////////
// Spell: Coldflame Damage

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 7]
```diff
@@ -1115,6 +1065,10 @@ class Coldflame : public SpellScript
...
```
Comment (schnek) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1068]: //////////////////////////////////////////////////////////////////////////////////////////
// Spell: Coldflame Bone Storm

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 8]
```diff
@@ -1073,6 +1019,10 @@ class BoneSpikeGraveyard : public SpellScript
         }        
 };
 
+//////////////////////////////////////////////////////////////////////////////////////////
```
Comment (schnek) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1022]: //////////////////////////////////////////////////////////////////////////////////////////
//

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 9]
```diff
@@ -988,6 +923,10 @@ class BoneStormDamage : public SpellScript
     }
 };
 
+//////////////////////////////////////////////////////////////////////////////////////////
```
Comment (schnek) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:926]: //////////////////////////////////////////////////////////////////////////////////////////
//

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 10]
```diff
@@ -970,6 +901,10 @@ class BoneStorm : public SpellScript
...
```
Comment (schnek) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:904]: //////////////////////////////////////////////////////////////////////////////////////////
//

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 11]
```diff
@@ -882,77 +882,8 @@ class BoneSpikeAI : public CreatureAIScript
...
-                std::set<uint32> mapsUsed;
-                GameObjectProperties const* goinfo = pInstance->AllianceTransporterAllianceShip->GetGameObjectProperties();
-
-                pInstance->AllianceTransporterAllianceShip->GenerateWaypoints(goinfo->raw.parameter_0);
-            }
-        }
-        GossipMenu::senGossipComplete(pPlayer);
-    }
-};
+// Spell: Bone Storm
```
Comment (schnek) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:885]: //////////////////////////////////////////////////////////////////////////////////////////
//

**Review by schnek — CHANGES_REQUESTED** [review]

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp (hunk 2 above)
Comment (aaron02) [reply src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1193]: You dont like it that way?

Comment (schnek) [reply src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1193]: stick to code style =) (although this is not the main thing in such a volume of changes)
https://ascemu.github.io/Wiki/docs/coding-standards/ Comments😉

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 12]
```diff
@@ -0,0 +1,2077 @@
...
+                if (!marrowgarAI)
+                    return;
+
+                uint8_t boneSpikeCount = uint8_t(aur->GetUnitCaster()->GetMapMgr()->pInstance->m_difficulty & 1 ? 3 : 1);
+                std::list<Unit*> targets;
+
+                targets.clear();
+
+                uint8_t target = 0;
+                for (target; target < boneSpikeCount; ++target)
```
Comment (Appled) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1218]: for (uint8_t target = 0; target < boneSpikeCount; ++target)
and remove declaration of target variable one line above

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 13]
```diff
@@ -0,0 +1,2077 @@
...
+        return SpellScriptEffectDamage::DAMAGE_FULL_RECALCULATION;
+    }
+};
+
+//////////////////////////////////////////////////////////////////////////////////////////
+/// Spell: Bone Spike Graveyard
+class BoneSpikeGraveyard : public SpellScript
+{
+    public:
+        virtual void onAuraApply(Aura* aur)
```
Comment (Appled) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1202]: void onAuraApply(...) override

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 14]
```diff
@@ -0,0 +1,2077 @@
...
+                emoteVector.clear();
+                emoteVector.push_back(SAY_MARR_BONESPIKE_1);// Stick Around
+                emoteVector.push_back(SAY_MARR_BONESPIKE_2);// The only Escape is Darkness 
+                emoteVector.push_back(SAY_MARR_BONESPIKE_3);// More Bones for the offering
+
+                marrowgarAI->sendRandomDBChatMessage(emoteVector);  
+            }
+        }
+
+        virtual SpellCastResult onCanCast(Spell* spell, uint32_t* parameter1, uint32_t* parameter2)
```
Comment (Appled) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1242]: SpellCastResult onCanCast(...) override

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 15]
```diff
@@ -0,0 +1,2077 @@
...
+        {          
+            if (Creature* marrowgar = static_cast<Creature*>(spell->getUnitCaster()))
+
+                if(GetRandomTargetNotMainTank(marrowgar))
+                    return SpellCastResult::SPELL_CAST_SUCCESS;
+
+            return SpellCastResult::SPELL_FAILED_SUCCESS;
```
Comment (Appled) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1249]: What do you want to do here? If you want the spell cast to fail, use a cast result like SPELL_FAILED_DONT_REPORT

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 16]
```diff
@@ -0,0 +1,2077 @@
...
+            return true;
+        }        
+};
+
+//////////////////////////////////////////////////////////////////////////////////////////
+/// Spell: Coldflame
+class Coldflame : public SpellScript
+{
+public:
+    virtual void filterEffectTargets(Spell* spell, uint8_t effectIndex, std::vector<uint64_t>* effectTargets)
```
Comment (Appled) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1296]: void filterEffectTargets(...) override

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 17]
```diff
@@ -0,0 +1,2077 @@
...
+
+            if (coldflametarget)
+            {
+                pCreature->GetScript()->SetCreatureData64(DATA_COLDFLAME_GUID, coldflametarget->getGuid());
+                effectTargets->push_back(coldflametarget->getGuid());
+            }
+        }
+    }
+
+    SpellScriptExecuteState beforeSpellEffect(Spell* spell, uint8_t effectIndex)
```
Comment (Appled) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1324]: Missing override tag

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 18]
```diff
@@ -0,0 +1,2077 @@
...
+        return SpellScriptExecuteState::EXECUTE_PREVENT;
+    }
+};
+
+//////////////////////////////////////////////////////////////////////////////////////////
+/// Spell: Coldflame Bone Storm
+class ColdflameBonestorm : public SpellScript
+{
+public:
+    SpellScriptExecuteState beforeSpellEffect(Spell* spell, uint8_t effectIndex)
```
Comment (Appled) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1340]: Missing override tag

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 19]
```diff
@@ -0,0 +1,2077 @@
...
+        return SpellScriptExecuteState::EXECUTE_PREVENT;
+    }
+};
+
+//////////////////////////////////////////////////////////////////////////////////////////
+/// Spell: Coldflame Damage
+class ColdflameDamage: public SpellScript
+{
+public:
+    virtual void filterEffectTargets(Spell* spell, uint8_t effectIndex, std::vector<uint64_t>* effectTargets)
```
Comment (Appled) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1357]: void filterEffectTargets(...) override

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 20]
```diff
@@ -0,0 +1,2077 @@
...
+        if (target->HasAura(SPELL_IMPALED))
+            return false;
+
+        if (target->GetDistance2dSq(spell->getUnitCaster()) > static_cast<float>(spell->getSpellInfo()->getEffectRadiusIndex(EFF_INDEX_0)) )
+            return false;
+
+        return true;
+    }
+
+    SpellScriptExecuteState beforeSpellEffect(Spell* spell, uint8_t effectIndex)
```
Comment (Appled) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1386]: Missing override tag

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 21]
```diff
@@ -0,0 +1,2077 @@
...
+            return SpellScriptEffectDamage::DAMAGE_DEFAULT;
+
+        if (!targetCount)
+            return SpellScriptEffectDamage::DAMAGE_DEFAULT;
+        
+        *dmg = float2int32(*dmg / (float)targetCount);
+        return SpellScriptEffectDamage::DAMAGE_FULL_RECALCULATION;
+    }
+
+    virtual SpellCastResult onCanCast(Spell* spell, uint32_t* parameter1, uint32_t* parameter2)
```
Comment (Appled) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1412]: SpellCastResult onCanCast(...) override

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 22]
```diff
@@ -0,0 +1,2077 @@
...
+    {
+        targetCount = 0;
+        static_cast<Creature*>(spell->getUnitCaster())->GetScript()->DoAction(ACTION_CLEAR_SPIKE_IMMUNITIES);
+
+        return SpellCastResult::SPELL_CAST_SUCCESS;
+    }
+
+    virtual void filterEffectTargets(Spell* spell, uint8_t effectIndex, std::vector<uint64_t>* effectTargets)
```
Comment (Appled) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1420]: void filterEffectTargets(...) override

File: src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp [hunk 23]
```diff
@@ -0,0 +1,2077 @@
...
+    {
+        if (effectIndex != EFF_INDEX_0)
+            return;
+
+        targetCount = effectTargets->size();        
+    }
+
+    virtual void afterSpellEffect(Spell* spell, uint8_t effIndex)
```
Comment (Appled) [inline src/scripts/InstanceScripts/Wotlk/Raid_IceCrownCitadel.cpp:1428]: void afterSpellEffect(...) override

File: src/world/Spell/Customization/HackFixes.cpp [hunk 24]
```diff
@@ -1150,6 +1164,15 @@ void SpellMgr::applyHackFixes()
 
     // Note: when applying spell hackfixes, please follow a template
 
+    //////////////////////////////////////////
+    // Lord Marrowgar                      //
+    ////////////////////////////////////////
+
+    sp = getMutableSpellInfo(69075); // Bone Storm (Lord Marrowgar)
+    if (sp != nullptr)
+        sp->setEffectImplicitTargetA(SPELL_TARGET_AREA, 0);
```
Comment (Appled) [inline src/world/Spell/Customization/HackFixes.cpp:1173]: Are you sure this is needed?

**Review by Appled — CHANGES_REQUESTED** [review]

File: src/world/Spell/Customization/HackFixes.cpp (hunk 24 above)
Comment (aaron02) [reply src/world/Spell/Customization/HackFixes.cpp:1173]: yeah dont worked without sadly

**Review by Appled — APPROVED** [review]
All good

**Comment by schnek** [comment]
```
public:
protected:
..
```
do not leave an empty line after them

and

```
method
{
public: // do not indent
    if (...
    {
       ....
    }
 }
```
😅

**Comment by aaron02** [comment]
> ```
> public:
> protected:
> ..
> ```
> 
> do not leave an empty line after them
> 
> and
> 
> ```
> method
> {
> public: // do not indent
>     if (...
>     {
>        ....
>     }
>  }
> ```
> 
> 😅

keywords
public, private, protected and friend should be indented.
Put single blank line after those keywords!
Any friend keywords should appear at the bottom of a class definition below any other code.

didnt i do it right?

**Comment by schnek** [comment]
> didnt i do it right?

no, just the opposite without a space.

**Comment by aaron02** [comment]
maybe i understand it wrong because english is not my main language but in https://ascemu.github.io/Wiki/docs/coding-standards/ it says 1 blank line

**Comment by schnek** [comment]
Yes, right. but CodeFactor thinks this is a bug (wiki I'll fix it)

**Comment by Appled** [comment]
> 
> 
> maybe i understand it wrong because english is not my main language but in https://ascemu.github.io/Wiki/docs/coding-standards/ it says 1 blank line

Yeah that's wrong in the wiki. Maybe @schnek can update it?

**Comment by schnek** [comment]
CodeFactor — 519 issues found // + new line
CodeFactor — 498 issues found // - new line

File: src/scripts/InstanceScripts/Tbc/Instance_OldHillsbradFoothills.cpp [hunk 25]
```diff
@@ -0,0 +1,272 @@
...
+    /*mgr->register_creature_script(CN_LIEUTENANT_DRAKE, &LieutenantDrakeAI::Create);
+    mgr->register_creature_script(CN_THRALL, &ThrallAI::Create);
+
+    Script* eGossip = new ErozionGossip();
+    mgr->register_creature_gossip(CN_EROZION, eGossip);
+    Script* bGossip = new BrazenGossip();
+    mgr->register_creature_gossip(CN_BRAZEN, bGossip);
+    Script* tGossip = new ThrallGossip();
+    mgr->register_creature_gossip(CN_THRALL, tGossip);*/
+}
```
[the remark this answers is not in the document]
Comment (aaron02) [reply src/scripts/InstanceScripts/Tbc/Instance_OldHillsbradFoothills.cpp:259]: whats missing here? in a later PR i will redo all the instances

**Pull request merged**